transcripts/chaincode-labs/2019-06-12-ethan-heilman-network-partitioning-attacks.mdwn

```diff
@@ -0,0 +1,135 @@
+
+...
+You could not use DNS for this, and I believe some other cryptocurrencies don't use DNS. My understanding of the reason that DNS is used is because you only talk to your local DNS server in many cases. If there were just one server that the first time a Bitcoin node booted up you requested from, then the access logs to that server would be everyone who runs Bitcoin core. Whereas DNS provides this layer of indirection that kind of hides who is asking queries. I don't know how good that actually is from a privacy standpoint, but even if you don't, the idea is that it does provide some benefit, there's not just a server out there that's a list of everyone who does it. That's why they use DNS verse a TLS to some server that gives you a JSON list of all the IP addresses.
+
+...the hard-coded IP addresses. How are those generated? And how many of them were online? My labmate was doing these really fun UDP fragmentation attacks for injecting traffic as an off-path attacker and doing DNS cache poisoning attacks you can just inject. You request it from some resolver, and then it's sent out a packet, and you do a fragmentation attack to inject your packet without being on-path or in-path. Because it's UDP, it doesn't have the same protections. So one idea to solve this is that you could try to authenticate the seeders. Some of the DNS response could actually be a signature on the other IP addresses. So you could encode a signature in the first 2 ipv4 addresses or the first couple ipv6 addresses. So you could make this signed data and sort of just shoehorn in sign data over DNS. Since you're already hard coding in the DNS seed, you could also just hardcore code in a pubkey for that DNS seed. Or you could look at turning off DNS cache poisoning, so you always make a new request. I haven't looked at the DNS code in a while, so maybe cache poisoning was turned off. I don't actually know that this isn't already fixed. But it's an interesting thing to look at. If you're asking DNS seeds, you should almost definitely not be using whatever cache you have locally. You should at least be going off to your local resolver. You could actually just call off the DNS through Bitcoin, but almost all applications rely on the OS to do that for you.
+
+# Conclusion and project ideas
+
+So I feel talked about all of this stuff. In general, some project ideas things that I think would be cool if people worked on was hardening the seeders against attacks, thinking about how to make on-path attacks harder. UDP has some interesting properties for on-path attacks. So off-path UDP is usually the easier to spoof packets because it's stateless, but if two parties send you UDP packets, you can't necessarily do these reset attacks where they just break the connection, and the person will drop packets. And then, incoming connection logic, is there any way to game this? How optimal is it? Is there some fine-tuning that would make it more secure?
```

Comment:
So I feel I talked about all of this stuff. In general, some **project ideas, things** that I think would be cool if people worked on **. One was** hardening the seeders against attacks...